packages/light.js/src/rpc/utils/createRpc.ts

```diff
@@ -3,99 +3,78 @@
-   *
-   * @param frequency - An array of frequency Observables.
-   * @example
-   * balanceOf$.setFrequency([onEverySecond$, onStartup$]); // Will fetch
-   * balance once on startup, and then every second.
-   */
-  setFrequency(frequency: Observable<any>[]) {
-    // TODO Check that frequency is well-formed
-
-    this.metadata.frequency = frequency;
-
-    // If necessary, we clear the memoize cache
-    if (typeof this.clear === 'function') {
-      this.clear();
-    }
-  }
-};
+import { createApiFromProvider, getApi } from '../../api';
+import { distinctValues } from '../../utils/operators';
+import { Metadata, RpcObservableOptions } from '../../types';
 
 /**
  * Add metadata to an RpcObservable, and transform it into a ReplaySubject(1).
+ * It's a currified function.
+ * Pure function version of {@link createRpc}.
  *
  * @ignore
  * @param metadata - The metadata to add.
- * @return - The original RpcObservable with patched metadata.
+ * @example
+ * createRpc(metadata)(options) returns a RpcObservable.
+ * createRpc(metadata)(options)(someArgs) returns an Observable.
  */
-const createRpc = <Source, Out>(metadata: Metadata<Source, Out>) => {
-  // rpc$ will hold the RpcObservable minus its metadata
-  const rpc$: RpcObservableWithoutMetadata<Source, Out> = (...args: any[]) => {
+const createRpcWithApi = memoizee(
+  <Source, Out>(metadata: Metadata<Source, Out>, api: any, ...args: any[]) => {
     // The source Observable can either be another RpcObservable (in the
     // `dependsOn` field), or anObservable built by merging all the
     // FrequencyObservables
     const source$ = metadata.dependsOn
-      ? metadata.dependsOn(...args)
-      : merge(...metadata.frequency);
-
-    // The last arguments is an options, if it's an object
-    // TODO What if we pass a single object as argument, which is not options?
-    const options: { withoutLoading?: boolean } =
-      args && args.length && isObject(args[args.length - 1]) ? args.pop() : {};
+      ? metadata.dependsOn(...args, { provider: api.provider })
+      : merge(...metadata.frequency.map(f => f({ provider: api.provider })));
 
-    // A RpcObservable is a source$ Observable, a single subject$ that
-    // subscribesthis source, and this subject$ multicasts the fired values to
-    // all Observers.
+    // A RpcObservable is: a source$ Observable, a single subject$ that
+    // subscribes to this source, and this subject$ multicasts the fired values
+    // to all Observers.
     const subject$ = new ReplaySubject<Out>(1);
 
     // The pipes to add
     const pipes: OperatorFunction<any, any>[] = [];
     if (metadata.pipes && isFunction(metadata.pipes)) {
-      pipes.push(...metadata.pipes(...args));
+      pipes.push(...metadata.pipes(api));
     }
-    pipes.push(multicast(() => subject$), refCount());
-    if (options.withoutLoading === true) {
-      pipes.push(withoutLoading());
+    pipes.push(multicast(() => subject$), refCount(), distinctValues());
```

Comment:
(we can probably use `distinctReplayRefCount` like in `createPubsubObservable` here)

Reply:
Totally correct.

I used to put the subject$ inside the RpcObservable's metadata, so we can track the number of subscribers, the current value etc, but removed all that.